Modules/_abc.c

```diff
@@ -0,0 +1,853 @@
+{
+    PyObject *data;
+    if (compute_abstract_methods(self) < 0) {
+        return NULL;
+    }
+
+    /* Set up inheritance registry. */
+    data = abc_data_new(&_abc_data_type, NULL, NULL);
+    if (data == NULL) {
+        return NULL;
+    }
+    if (_PyObject_SetAttrId(self, &PyId__abc_impl, data) < 0) {
+        Py_DECREF(data);
+        return NULL;
+    }
+    Py_DECREF(data);
+    Py_RETURN_NONE;
+}
+
+/*[clinic input]
+_abc._abc_register
+
+    self: object
+    subclass: object
+    /
+
+Internal ABC helper for subclasss registration. Should be never used outside abc module
+[clinic start generated code]*/
+
+static PyObject *
+_abc__abc_register_impl(PyObject *module, PyObject *self, PyObject *subclass)
+/*[clinic end generated code: output=7851e7668c963524 input=4b94c86e77e9e901]*/
+{
+    if (!PyType_Check(subclass)) {
+        PyErr_SetString(PyExc_TypeError, "Can only register classes");
+        return NULL;
+    }
+    int result = PyObject_IsSubclass(subclass, self);
+    if (result > 0) {
+        Py_INCREF(subclass);
+        return subclass;  /* Already a subclass. */
+    }
+    if (result < 0) {
+        return NULL;
+    }
+    /* Subtle: test for cycles *after* testing for "already a subclass";
+       this means we allow X.register(X) and interpret it as a no-op. */
+    result = PyObject_IsSubclass(self, subclass);
+    if (result > 0) {
+        /* This would create a cycle, which is bad for the algorithm below. */
+        PyErr_SetString(PyExc_RuntimeError, "Refusing to create an inheritance cycle");
+        return NULL;
+    }
+    if (result < 0) {
+        return NULL;
+    }
+    _abc_data *impl = _get_impl(self);
+    if (impl == NULL) {
+        return NULL;
+    }
+    if (_add_to_weak_set(&impl->_abc_registry, subclass) < 0) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    Py_DECREF(impl);
+
+    /* Invalidate negative cache */
+    PyObject *one = PyLong_FromLong(1);
+    if (one == NULL) {
+        return NULL;
+    }
+    PyObject *next_version = PyNumber_Add(abc_invalidation_counter, one);
+    Py_DECREF(one);
+    if (next_version == NULL) {
+        return NULL;
+    }
+    Py_SETREF(abc_invalidation_counter, next_version);
+
+    Py_INCREF(subclass);
+    return subclass;
+}
+
+
+/*[clinic input]
+_abc._abc_instancecheck
+
+    self: object
+    instance: object
+    /
+
+Internal ABC helper for instance checks. Should be never used outside abc module
+[clinic start generated code]*/
+
+static PyObject *
+_abc__abc_instancecheck_impl(PyObject *module, PyObject *self,
+                             PyObject *instance)
+/*[clinic end generated code: output=b8b5148f63b6b56f input=5b15699272795a93]*/
+{
+    PyObject *subtype, *result = NULL, *subclass = NULL;
+    _abc_data *impl = _get_impl(self);
+    if (impl == NULL) {
+        return NULL;
+    }
+
+    subclass = _PyObject_GetAttrId(instance, &PyId___class__);
```

Comment:
Check for `subclass == NULL`.

Reply:
But I thought that it is guaranteed to never fail. Trying to delete `.__class__` attribute raises `TypeError`.

Reply:
What if the class defines its own `__class__` descriptor?